go/signatures_test.go

```diff
@@ -0,0 +1,310 @@
+	pk, err := PublicKeyFromPEM([]byte(sigTestRSAPublicKeyPEM))
+	if err != nil {
+		t.Fatalf("Failed to parse sigTestRSAPublicKey: %v", err)
+	}
+	return pk
+}
+
+func sigTestCertLogEntry(t *testing.T) LogEntry {
+	return LogEntry{
+		Index: 0,
+		Leaf: MerkleTreeLeaf{
+			Version:  V1,
+			LeafType: TimestampedEntryLeafType,
+			TimestampedEntry: TimestampedEntry{
+				Timestamp: sigTestSCTTimestamp,
+				EntryType: X509LogEntryType,
+				X509Entry: mustDehex(t, sigTestDERCertString),
+			},
+		},
+	}
+}
+
+func mustCreateSignatureVerifier(t *testing.T, pk crypto.PublicKey) SignatureVerifier {
+	sv, err := NewSignatureVerifier(pk)
+	if err != nil {
+		t.Fatalf("Failed to create SignatureVerifier: %v", err)
+	}
+	return *sv
+}
+
+func TestVerifySCTSignatureEC(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTEC(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureRSA(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestRSAPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForMismatchedSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err == nil {
+		t.Fatalf("Sucessfully verified with mismatched signature algorithm")
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForUnknownSignatureAlgorith(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedSignatureAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported signature algorithm")
+	}
+}
+
+func TestVerifySCTSignatureFailsForUnknownHashAlgorith(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedHashAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported hash algorithm")
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectInput(t *testing.T, sv SignatureVerifier) {
+	entry := sigTestCertLogEntry(t)
+	entry.Leaf.TimestampedEntry.X509Entry[mrand.Intn(len(entry.Leaf.TimestampedEntry.X509Entry))] ^= byte(mrand.Intn(255) + 1)
+	if err := sv.VerifySCTSignature(sigTestSCTEC(t), entry); err == nil {
+		t.Fatal("Incorrectly verfied signature over corrupted input data.")
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectSignature(t *testing.T, sv SignatureVerifier) {
+	sct := sigTestSCTEC(t)
+	sct.Signature.Signature[mrand.Intn(len(sct.Signature.Signature))] ^= byte(mrand.Intn(255) + 1)
```

Comment:
Technically, a signature could have bytes you can change (e.g. unsigned metadata) but we should worry about that if it ever happens. :-)